tanglegram/tangle.py

```diff
@@ -754,62 +651,28 @@ def leaf_order(link, labels=None, as_dict=True):
             return leafs_ix
 
 
-def entanglement(lindex1, lindex2, L=1.5):
-    """Calculage average displacement of leafs in dendogram 1 and 2.
-
-    Entanglement is a measure between 1 (full entanglement) and 0 (no
-    entanglement). Ignores leafs that aren't present in both dendrograms.
-
-    Parameters
-    ----------
-    lindex1,lindex2 :       dict
-                            Dictionaries mapping the labels of two dendrograms
-                            to their indices.
-    L :                     any positive number
-                            Distance norm to use for measuring the distance
-                            between the two trees. Can be any positive number,
-                            often one will want to use 0, 1, 1.5 or 2:
-                            ``sum(abs(x-y)^L)``.
-
+def entanglement(link1, link2):
+    """ Entanglement is measured by giving the left tree’s labels the values of 1 till tree size,
+    matching these numbers with the right tree, and then dividing the sum of the square difference
+    between elements of these two vectors (sum(abs(x-y)^2)) by that number of the "worst case" entanglement
+    (e.g: when the right tree is the complete reverse of the left tree).
+    Skips leafs that aren't present in both dendrograms.
     """
-    assert isinstance(lindex1, dict)
-    assert isinstance(lindex2, dict)
+    lindex1 = leaves_list(link1)
+    lindex2 = leaves_list(link2)
 
     exist_in_both = list(set(lindex1) & set(lindex2))
+    ix = np.arange(max(len(lindex1), len(lindex2)))
 
     if not exist_in_both:
         raise ValueError('Not a single matching label in both dendrograms.')
 
-    # Absolute distance
-    dist = np.array([lindex1[l] - lindex2[l] for l in exist_in_both])
-    dist = np.abs(dist)
-    # Absolute entanglement
-    ent = np.sum(dist ** L)
-
-    # Worst case
-    ix = np.arange(max(len(lindex1), len(lindex2)))
-    worst = np.sum(np.abs(ix - ix[::-1]) ** L)
-
-    # Normalized entanglemtn
-    return ent / worst
-
-
-if __name__ == '__main__':
-    labelsA = ['A', 'B', 'C', 'D']
-    labelsB = ['B', 'A', 'C', 'D']
-    data = [[ 0,  .1,  .4, .3],
-            [.1,   0,  .5, .6],
-            [ .4, .5,   0, .2],
-            [ .3, .6,  .2,  0]]
-
-    mat1 = pd.DataFrame(data,
-                        columns=labelsA,
-                        index=labelsA)
-
-    mat2 = pd.DataFrame(data,
-                        columns=labelsB,
-                        index=labelsB)
+    # Mapping the "number" (1 til tree size) in the left tree with the right tree
+    matching_leaf_vector = np.zeros(max(len(lindex1), len(lindex2)))
+    for i in lindex2:
```

Comment:
I haven't tested it properly but this for loop can't possibly be faster than the previous array-based solution. Could you elaborate a bit on what the advantage of doing it this way is?

Reply:
The previous array-based solution is actually not correct. Here we want to match the "number" (1 til tree size) in the left tree with the right tree and then calculate the difference between these numbers in two trees, not to compute the difference between indices. Do you apply such matching with "dict" which I do not understand honestly?

The old calculation leads to different results compared to R language.

Reply:
I disagree re the existing solution being incorrect - it might not yield exactly the same results as in R but certainly does the same job. Using `{label: index}` dicts was necessary for scenarios where labels in both dendrograms don't match up 1:1.

Reply:
I agree it is necessary to use such dictionary. But I am not sure it matches the numbers (from 1 to tree size) in the left tree with the right tree.  What I wanted can be illustrated in the following example:
Left tree: A D E C F
Right tree: C D A F E
Giving objects in the left tree numbers from 1 to tree size yields: 1, 2, 3, 4, 5
Matching these numbers with the right tree: 4, 2, 1, 5, 3

Reply:
That's pretty much what the existing function does with the dict.

Reply:
Then that's my bad not realizing it.